sources/updater/update_metadata/src/lib.rs

```diff
@@ -419,35 +450,207 @@ pub fn load_manifest<T: tough::Transport>(repository: &tough::Repository<T>) ->
+            waves,
+            images: Images {
+                boot: String::from("boot"),
+                root: String::from("root"),
+                hash: String::from("hash"),
+            },
+        };
+
+        for seed in (100..500).step_by(100) {
+            assert!(
+                !update.update_ready(seed, time),
+                "seed: {}, time: {}, wave start time: {}, wave start seed: {}, {} milliseconds hasn't passed yet",
+                seed,
+                time,
+                time,
+                0,
+                seed
+            );
+            assert!(
+                update.update_ready(seed, time + Duration::milliseconds(seed as i64)),
+                "seed: {}, time: {}, wave start time: {}, wave start seed: {}, update should be ready",
+                seed,
+                time + Duration::milliseconds(100),
+                time,
+                0,
+            );
+        }
+
+        // Test multiple waves
+        update.waves.clear();
+        let mut seed = 148;
+        assert!(
+            update.update_ready(seed, time),
+            "No waves specified but no update"
+        );
 
-#[test]
-fn test_migrations_backward() {
-    // The same manifest as `test_migrations_forward` but this time we will migrate backward.
-    let path = "./tests/data/migrations.json";
-    let manifest: Manifest = serde_json::from_reader(File::open(path).unwrap()).unwrap();
-    let from = Version::parse("1.5.0").unwrap();
-    let to = Version::parse("1.0.0").unwrap();
-    let targets = find_migrations(&from, &to, &manifest).unwrap();
-
-    assert!(targets.len() == 3);
-    let mut i = targets.iter();
-    assert!(i.next().unwrap() == "migration_1.5.0_shortcut");
-    assert!(i.next().unwrap() == "migration_1.1.0_b");
-    assert!(i.next().unwrap() == "migration_1.1.0_a");
+        update.waves.insert(0, time);
+        // First wave ends 200 milliseconds into the update and has seeds 0 - 50
+        update.waves.insert(50, time + Duration::milliseconds(200));
+        // Second wave ends 1024 milliseconds into the update and has seeds 50 - 100
+        update
+            .waves
+            .insert(100, time + Duration::milliseconds(1024));
+        // Third wave ends 4096 milliseconds into the update and has seeds 100 - 1024
+        update
+            .waves
+            .insert(1024, time + Duration::milliseconds(4096));
+
+        for duration in (0..200).step_by(10) {
+            assert!(
+                !update.update_ready(seed, time + Duration::milliseconds(duration)),
+                "seed should not part of first wave",
+            );
+        }
+
+        for duration in (200..1024).step_by(4) {
+            assert!(
+                !update.update_ready(seed, time + Duration::milliseconds(duration)),
+                "seed should not part of second wave",
+            );
+        }
+
+        for duration in (1024..((4096 - 1024) / (1024 - 100) * seed)).step_by(4) {
```

Comment:
Can we pull this out into a variable and explain what it represents?  Same below?  Clarifying exactly when the seed should become available would be helpful.